apps/web/src/components/code-editor/components/editor/index.tsx

```diff
@@ -0,0 +1,88 @@
+import React, { forwardRef, useCallback, useMemo } from 'react';
+import cls from 'classnames';
+import CodeMirror, { type ReactCodeMirrorRef } from '@uiw/react-codemirror';
+// support language
+import { yaml } from '@codemirror/lang-yaml';
+import { json } from '@codemirror/lang-json';
+import { java } from '@codemirror/lang-java';
+import { python } from '@codemirror/lang-python';
+import { javascript } from '@codemirror/lang-javascript';
+import { markdown } from '@codemirror/lang-markdown';
+// extension language
+import { StreamLanguage } from '@codemirror/language';
+import { groovy } from '@codemirror/legacy-modes/mode/groovy';
+
+import { useEditorTheme } from '../../hooks';
+import type { EditorContentProps, EditorSupportLang } from '../../types';
+import './style.less';
+
+interface IProps extends EditorContentProps {
+    editorLang?: EditorSupportLang;
+    editorValue: string;
+    setEditorValue: (value: string) => void;
+}
+export const CodeEditorContent = forwardRef<ReactCodeMirrorRef, IProps>((props, ref) => {
+    const {
+        showLineNumber = true,
+        showFold = true,
+        editable = true,
+        readOnly = false,
+        editorLang,
+        editorValue,
+        fontSize,
+        setEditorValue,
+        ...rest
+    } = props;
+    const { editorTheme } = useEditorTheme({ fontSize });
+
+    /** editor input change callback */
+    const onInputChange = useCallback(
+        (value: string) => {
+            setEditorValue(value);
+        },
+        [setEditorValue],
+    );
+
+    /** Select the corresponding extension based on the language */
+    const extensions = useMemo(() => {
+        switch (editorLang) {
+            case 'yaml':
+                return [yaml()];
+            case 'json':
+                return [json()];
+            case 'mvel':
+                // The mvel language is similar to java, which is used here for highlighting
+                return [java()];
+            case 'python':
+                return [python()];
+            case 'javascript':
+                return [javascript()];
+            case 'groovy':
+                return [StreamLanguage.define(groovy)];
+            case 'markdown':
+                return [markdown()];
+            case 'text':
+            default:
+                return [];
+        }
+    }, [editorLang]);
+
+    return (
+        <CodeMirror
+            {...rest}
+            ref={ref}
+            className={cls('ms-code-editor-content', {
+                [`ms-editor__lineNumbers--hide`]: !showLineNumber,
+                [`ms-editor__foldGutter--hide`]: !showFold,
+            })}
+            value={editorValue}
+            extensions={extensions}
+            onChange={onInputChange}
```

Comment:
`value`, `onChange` property is occupied, how can the parent component control or listen to the data change?

Reply:
The component uses the `useControllableValue` api of ahooks, which is controlled externally when a `value`, `onChange` is passed in externally, otherwise the internal state manages the state.

The advantage of using useControllableValue is that the component can use the same api for operation assignment, reducing the operation burden